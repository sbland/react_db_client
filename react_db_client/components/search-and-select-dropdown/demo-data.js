export const demoHeadingsData = [
    {
      type: 'text',
      uid: 'uid',
      label: 'UID',
    },
    {
      type: 'text',
      uid: 'type',
      label: 'Type',
    },
    {
      type: 'text',
      uid: 'name',
      label: 'Name',
    },
  ];

  export const demoPreviewHeadingsData = [
    {
      type: 'text',
      uid: 'uid',
      label: 'UID',
    },
    {
      type: 'text',
      uid: 'type',
      label: 'Type',
    },
    {
      type: 'text',
      uid: 'name',
      label: 'Name',
    },
    {
      type: 'textLong',
      uid: 'description',
      label: 'Description',
    },
  ];

  export const demoResultData = [
    {
      uid: '1',
      label: 'Result 1',
      type: 'text',
      name: 'This thing 1',
    },
    {
      uid: '2',
      label: 'Result 2',
      type: 'text',
      name: 'This thing 2',
    },
    {
      uid: '3',
      label: 'Result 3',
      type: 'text',
      name: 'This thing 3',
    },
    {
      uid: '4',
      label: 'Result 4',
      type: 'text',
      name: 'This thing 4',
    },
    {
      uid: '5',
      label: 'Result 5',
      type: 'text',
      name: 'This thing 5',
    },
    {
      uid: '6',
      label: 'Result 6',
      type: 'text',
      name: 'This thing 6',
    },
  ];

  export const demoResultsDataMany = Array.from(Array(30).keys()).map(
    (i) => demoResultData[i % demoResultData.length]
  );
