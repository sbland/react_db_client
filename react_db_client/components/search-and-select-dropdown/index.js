export { SearchAndSelectDropdown } from './search-and-select-dropdown';
