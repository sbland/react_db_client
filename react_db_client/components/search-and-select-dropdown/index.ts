export * from './search-and-select-dropdown';
