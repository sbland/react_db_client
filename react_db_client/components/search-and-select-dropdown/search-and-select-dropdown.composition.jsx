/* eslint-disable react/prop-types */
import React, { useState } from 'react';
import { CompositionWrapDefault } from '@react_db_client/helpers.composition-wraps';

// eslint-disable-next-line import/no-extraneous-dependencies
import { SearchAndSelectDropdown } from './search-and-select-dropdown';
import { demoResultData } from './demo-data';

const defaultProps = {
  searchFunction: async () =>
    new Promise((resolve) => setTimeout(() => resolve(demoResultData), 500)),
  // handleSelect: (id) => alert(`Selected: ${id}`),
  handleSelect: (id) => {},
  labelField: 'label',
  debug: true,
  searchFieldTargetField: 'label',
};

export const DemoData = () => {
  const [allowEmptySearch, setAllowEmptySearch] = useState(false);
  const [searchDelay, setSearchDelay] = useState(500);
  const props = { ...defaultProps, allowEmptySearch, searchDelay };
  return (
    <div className="">
      <div className="">
        <button
          type="button"
          className={allowEmptySearch ? 'button-two' : 'button-one'}
          onClick={() => setAllowEmptySearch((prev) => !prev)}
        >
          Allow empty Search
        </button>
        <label>Search Delay</label>
        <input
          type="number"
          name="searchDelay"
          onChange={(e) => setSearchDelay(e.target.value)}
          value={searchDelay}
        />
      </div>
      <CompositionWrapDefault height="4rem" width="8rem">
        <SearchAndSelectDropdown {...props} style={{ background: 'red' }} />
      </CompositionWrapDefault>
    </div>
  );
};

export const DemoDataMultipleLabels = () => {
  const props = { ...defaultProps, labelField: ['label', 'uid'] };
  return (
    <div>
      <SearchAndSelectDropdown {...props} allowEmptySearch />
    </div>
  );
};
export const DemoDataAltLabel = () => {
  const results = demoResultData.map((item) => ({
    ...item,
    label: null,
    name: item.label,
  }));
  const props = {
    ...defaultProps,
    labelField: 'name',
    allowMultiple: true,
    searchFunction: async () =>
      new Promise((resolve) => setTimeout(() => resolve(demoResultData), 500)),
  };
  return (
    <div>
      <SearchAndSelectDropdown {...props} allowEmptySearch />
    </div>
  );
};

export const DemoDataAllowEmptyInstant = () => {
  const props = { ...defaultProps };
  return (
    <div>
      <SearchAndSelectDropdown
        {...props}
        allowEmptySearch
        searchFunction={async () => demoResultData}
        searchDelay={0}
      />
    </div>
  );
};



export const DemoDataSetInitial = () => {
  const props = { ...defaultProps };
  return (
    <div>
      <SearchAndSelectDropdown
        {...props}
        initialValue="hello"
        allowEmptySearch
        searchFunction={async () => demoResultData}
        searchDelay={0}
      />
    </div>
  );
};
