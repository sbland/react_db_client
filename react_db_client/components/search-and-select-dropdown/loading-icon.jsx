import React from 'react';
import PropTypes from 'prop-types';
import { Emoji } from '@react_db_client/components.emoji';

export const LoadingIcon = ({ isLoading }) => {
  return (
    <>
      {isLoading && (
        <Emoji emoj="\/" label="loading" className="loadingIcon"/>
        // <Emoji emoj="âŒ›" label="loading" className="loadingIcon"/>
        // <img
        //   className="loadingIcon"
        //   src="/images/icons/loadingIcon.png"
        //   alt="loading"
        //   height="100%"
        // />
      )}
    </>
  );
};

LoadingIcon.propTypes = {
  isLoading: PropTypes.bool.isRequired,
};
