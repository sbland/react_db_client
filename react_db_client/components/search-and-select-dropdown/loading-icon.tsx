import React from 'react';
import PropTypes from 'prop-types';
import { Emoji } from '@react_db_client/components.emoji';
import { LoadingIconStyle } from './styles';

export interface ILoadingIconProps {
  isLoading: boolean;
}

export const LoadingIcon = ({ isLoading }: ILoadingIconProps) => {
  return (
    <>
      {isLoading && (
        <LoadingIconStyle data-testid="loadingIcon">
          <Emoji emoj="\/" label="loading" className="loadingIcon" />
        </LoadingIconStyle>
        // <Emoji emoj="âŒ›" label="loading" className="loadingIcon"/>
        // <img
        //   className="loadingIcon"
        //   src="/images/icons/loadingIcon.png"
        //   alt="loading"
        //   height="100%"
        // />
      )}
    </>
  );
};

LoadingIcon.propTypes = {
  isLoading: PropTypes.bool.isRequired,
};
