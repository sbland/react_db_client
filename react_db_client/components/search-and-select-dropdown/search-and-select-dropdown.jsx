import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import {
  FilterObjectClass,
  filterTypes,
  comparisons,
} from '@samnbuk/react_db_client.constants.client-types';
import { useAsyncRequest } from '@samnbuk/react_db_client.async-hooks.use-async-request';
import { CustomSelectDropdown } from '@samnbuk/react_db_client.components.custom-select-dropdown';

import { LoadingIcon } from './loading-icon';

import './style.scss';

/**
 * Search and Select Dropdown Component
 * Dropdown selection with async data load

 */
export const SearchAndSelectDropdown = ({
  searchFunction,
  handleSelect,
  intitialValue,
  searchFieldTargetField, // the target field that the search string applies to
  labelField, // The field in the returned data to use as the label
  className,
  searchFieldPlaceholder,
  allowEmptySearch,
  searchDelay,
  valid,
}) => {
  // TODO: Provide default search function
  const [searchValue, setSearchValue] = useState(() => intitialValue);
  const [isFocused, setIsFocused] = useState(false);
  const [loading, setLoading] = useState(false);
  const [showResults, setShowResults] = useState(false);
  const firstItemRef = useRef(null);
  const searchFieldRef = useRef(null);
  const searchTimeout = useRef(null);

  const goBackToSearchField = () => searchFieldRef.current.select();
  const [results, setResults] = useState([]);

  const searchCallback = useCallback((resultsNew) => {
    setResults(resultsNew);
    setLoading(false);
  }, []);

  const { reload } = useAsyncRequest({
    args: [],
    callFn: searchFunction,
    callOnInit: false,
    callback: searchCallback,
  });

  const handleItemSelect = handleSelect;

  const onSearchFieldChange = (e) => {
    setSearchValue(e.target.value);
    setResults([]);
    setShowResults(false);
    setLoading(true);
  };

  useEffect(() => {
    /* Set results to [] if we invalid search value */
    if (isFocused && !searchValue && !allowEmptySearch) setResults([]);
  }, [isFocused, searchValue, allowEmptySearch]);

  useEffect(() => {
    /* Set results to [] if loading */
    if (isFocused && loading) {
      setResults([]);
      setShowResults(false);
    }
  }, [isFocused, loading]);

  useEffect(() => {
    /* Show results if focused */
    if (isFocused && (searchValue || allowEmptySearch) && results?.length > 0) setShowResults(true);
  }, [isFocused, searchValue, allowEmptySearch, results]);

  useEffect(() => {
    /* Repeat search */
    if ((searchValue || allowEmptySearch) && isFocused) {
      const searchFilter = new FilterObjectClass({
        uid: 'search',
        field: searchFieldTargetField,
        value: searchValue,
        operator: comparisons.contains,
        type: filterTypes.text,
      });
      clearTimeout(searchTimeout.current);
      const args = searchValue ? [searchFilter] : [];
      searchTimeout.current = setTimeout(() => reload([args]), searchDelay);
    }
    return () => {
      clearTimeout(searchTimeout.current);
    };
  }, [searchValue, reload, searchFieldTargetField, allowEmptySearch, isFocused, searchDelay]);

  const mappedResults = useMemo(() => {
    if (Array.isArray(labelField)) {
      return Array.isArray(results)
        ? results.map((r) => ({
            uid: r.uid,
            label: labelField.map((L) => r[L]).join(' | '),
          }))
        : [];
    }
    return Array.isArray(results)
      ? results.map((r) => ({
          uid: r.uid,
          label: r[labelField],
        }))
      : [];
  }, [labelField, results]);

  const handleListItemSelect = (selectedId) => {
    if (!loading) {
      const selectedData = results.find((r) => r.uid == selectedId);
      handleItemSelect(selectedId, selectedData);
      setSearchValue(selectedData[labelField]);
      setIsFocused(false);
      setShowResults(false);
      goBackToSearchField();
    }
  };

  const handleCancel = () => {
    setShowResults(false);
    setIsFocused(false);
  };

  const handleEnterSearchField = useCallback(() => {
    setIsFocused(true);
    // if(searchValue || allowEmptySearch) setLoadin
  }, []);

  const handleInputKeyDown = (e) => {
    /* When in search field */
    if (e.key === 'Enter') {
      e.preventDefault();
      setShowResults(true);
    }
    if (e.key === 'Escape' || e.key === 'Tab') {
      handleCancel();
    }
    if (e.key === 'ArrowDown') {
      if (mappedResults && mappedResults.length > 0) {
        e.preventDefault();
        if (showResults && !loading) {
          firstItemRef.current.focus();
        } else {
          setShowResults(true);
        }
      }
    }
  };

  const classNames = [className, 'sas_drop_wrap', valid ? '' : 'invalid']
    .filter((a) => a)
    .join(' ');
  return (
    <div className={classNames}>
      <div className="searchFieldWrap">
        <input
          className="searchField"
          style={{ flexGrow: 1 }}
          type="text"
          placeholder={searchFieldPlaceholder}
          value={searchValue || ''}
          onChange={onSearchFieldChange}
          onFocus={handleEnterSearchField}
          onBlur={() => setIsFocused(false)}
          onKeyDown={handleInputKeyDown}
          ref={searchFieldRef}
        />
        {!loading && (
          <button
            type="button"
            className="dropdownBtn"
            onClick={() => setShowResults((prev) => !prev)}
          >
            \/
          </button>
        )}
        <div className="sas_drop_loadingWrap">
          <LoadingIcon isLoading={loading} />
        </div>
      </div>
      <CustomSelectDropdown
        options={mappedResults}
        isOpen={showResults}
        handleSelect={(uid) => handleListItemSelect(uid)}
        handleClose={() => setShowResults(false)}
        firstItemRef={firstItemRef}
        goBackToSearchField={goBackToSearchField}
      />
    </div>
  );
};

const selectionShape = {};

SearchAndSelectDropdown.propTypes = {
  searchFunction: PropTypes.func.isRequired,
  handleSelect: PropTypes.func.isRequired,
  selectionOverride: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.shape(selectionShape)),
    PropTypes.shape(selectionShape),
  ]),
  allowMultiple: PropTypes.bool,
  returnFieldOnSelect: PropTypes.string,
  searchFieldTargetField: PropTypes.string,
  labelField: PropTypes.oneOfType([PropTypes.string, PropTypes.arrayOf(PropTypes.string)])
    .isRequired,
  className: PropTypes.string,
  searchFieldPlaceholder: PropTypes.string,
  allowEmptySearch: PropTypes.bool,
  searchDelay: PropTypes.number,
};

SearchAndSelectDropdown.defaultProps = {
  selectionOverride: null,
  allowMultiple: false,
  returnFieldOnSelect: 'uid',
  searchFieldTargetField: 'label',
  className: '',
  searchFieldPlaceholder: 'search...',
  allowEmptySearch: false,
  searchDelay: 500,
};
