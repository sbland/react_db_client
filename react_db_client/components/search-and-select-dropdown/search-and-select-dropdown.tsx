import React, {
  useLayoutEffect,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import PropTypes from 'prop-types';
import {
  FilterObjectClass,
  EComparisons,
} from '@react_db_client/constants.client-types';
import {
  AsyncRequestError,
  useAsyncRequest,
} from '@react_db_client/async-hooks.use-async-request';
import { CustomSelectDropdown } from '@react_db_client/components.custom-select-dropdown';
import { useCombinedRefs } from '@react_db_client/hooks.use-combined-ref';
import { EFilterType } from '@react_db_client/constants.client-types';

import { LoadingIcon } from './loading-icon';
import {
  DropdownBtn,
  SasDropLoadingWrap,
  SasDropWrap,
  SearchFieldWrapStyle,
} from './styles';

export interface IItem {
  uid: string | number;
  label: string;
}

export type SelectFn<T extends IItem> = (selectedData: T) => void;

export interface ISearchAndSelectDropdownProps<Item extends IItem>
  extends React.HTMLProps<HTMLInputElement> {
  searchFunction: (filters?: FilterObjectClass[]) => Promise<Item[]>;
  handleSelect: (selectedData: Item) => void;
  debug?: boolean;
  initialValue?: null | string | Item;
  searchFieldTargetField?: string;
  labelField?: string | string[];
  className?: string;
  searchFieldPlaceholder?: string;
  allowEmptySearch?: boolean;
  autoFocusOnFirstItem?: boolean;
  searchDelay?: number;
  allowMultiple?: boolean;
  valid?: boolean;
  searchFieldRef?: React.MutableRefObject<HTMLInputElement | null> | null;
}

// TODO: Provide default search function

/**
 * Search and Select Dropdown Component
 * Dropdown selection with async data load
 */
export const SearchAndSelectDropdown = <Item extends IItem>(
  props: ISearchAndSelectDropdownProps<Item>
) => {
  const {
    searchFunction,
    handleSelect,
    initialValue,
    debug,
    searchFieldTargetField = 'field',
    labelField = 'label',
    className,
    searchFieldPlaceholder,
    allowEmptySearch,
    searchDelay,
    allowMultiple,
    autoFocusOnFirstItem,
    valid = true,
    style,
    searchFieldRef: searchFieldRefFromParent,
    ...additionalProps
  } = props;
  if (allowMultiple) throw Error('Multiple Selection Is Not Implemented');
  const [searchValue, setSearchValue] = useState<string | number | null>(null);
  const [isFocused, setIsFocused] = useState(false);
  const [loading, setLoading] = useState(false);
  const [showResults, setShowResults] = useState(false);
  const [waitForInput, setWaitForInput] = useState(false);
  const firstItemRef = useRef<HTMLElement | null>(null);
  const searchFieldRef = useRef<HTMLInputElement>(null);
  const searchFieldRefsCombined = useCombinedRefs(
    searchFieldRefFromParent,
    searchFieldRef
  );
  const searchTimeout: React.MutableRefObject<ReturnType<
    typeof setTimeout
  > | null> = useRef(null);

  const [results, setResults] = useState<Item[]>([]);
  const isTypingTypingRef: React.MutableRefObject<ReturnType<
    typeof setTimeout
  > | null> = React.useRef(null);
  const [isTyping, setIsTyping] = React.useState(false);

  React.useEffect(() => {
    setSearchValue(
      initialValue === null ||
        initialValue === undefined ||
        typeof initialValue !== 'object'
        ? initialValue || ''
        : initialValue[searchFieldTargetField]
    );
  }, [initialValue, searchFieldTargetField]);

  const searchCallback = useCallback((resultsNew) => {
    // console.info(resultsNew)
    setResults(resultsNew);
    setLoading(false);
  }, []);

  const searchErrorCallback = useCallback((error: AsyncRequestError) => {
    setResults([]);
    setLoading(false);
    // TODO: Handle errors
  }, []);

  const { reload } = useAsyncRequest({
    args: [],
    callFn: searchFunction,
    callOnInit: false,
    callback: searchCallback,
    errorCallback: searchErrorCallback,
  });

  const onSearchFieldChange = (e) => {
    setSearchValue(e.target.value);
    setResults([]);
    setIsTyping(true);
    setShowResults(false);
    setLoading(true);
    setWaitForInput(false);
  };

  useEffect(() => {
    if (isTyping) {
      isTypingTypingRef.current = setTimeout(() => {
        setIsTyping(false);
      }, 100);
    }
    return () => {
      if (isTypingTypingRef?.current) clearTimeout(isTypingTypingRef.current);
    };
  }, [isTyping]);

  const search = useCallback(() => {
    setResults([]);
    setLoading(true);
    setShowResults(false);
    const searchFilter = new FilterObjectClass({
      uid: 'search',
      field: searchFieldTargetField,
      value: searchValue,
      operator: EComparisons.CONTAINS,
      type: EFilterType.text,
    });
    if (searchTimeout.current) {
      clearTimeout(searchTimeout.current as NodeJS.Timeout);
    }
    const args = searchValue ? [searchFilter] : [];
    searchTimeout.current = setTimeout(
      () => reload([args]),
      searchDelay
    ) as unknown as NodeJS.Timeout;
  }, [searchFieldTargetField, searchValue, searchDelay, searchTimeout]);

  useEffect(() => {
    /* Set results to [] if invalid search value */
    if (isFocused && !searchValue && !allowEmptySearch) {
      setResults([]);
      setLoading(false);
    }
  }, [isFocused, searchValue, allowEmptySearch]);

  useEffect(() => {
    /* Set results to [] if loading */
    if (isFocused && loading) {
      setResults([]);
      setShowResults(false);
    }
  }, [isFocused, loading]);

  useEffect(() => {
    /* Show results if focused */
    if (
      !isTyping &&
      !waitForInput &&
      isFocused &&
      (searchValue || allowEmptySearch) &&
      results?.length > 0
    )
      setShowResults(true);
  }, [
    isTyping,
    isFocused,
    searchValue,
    allowEmptySearch,
    results,
    waitForInput,
  ]);

  useLayoutEffect(() => {
    if (autoFocusOnFirstItem && mappedResults && firstItemRef?.current) {
      firstItemRef.current.focus();
    }
  }, [autoFocusOnFirstItem, showResults]);

  useEffect(() => {
    /* Repeat search when focused */
    if (!waitForInput && (searchValue || allowEmptySearch) && isFocused) {
      search();
    }
    return () => {
      clearTimeout((searchTimeout.current as NodeJS.Timeout) || undefined);
    };
  }, [
    searchValue,
    reload,
    searchFieldTargetField,
    allowEmptySearch,
    isFocused,
    searchDelay,
    waitForInput,
  ]);

  const mappedResults = useMemo(() => {
    if (Array.isArray(labelField)) {
      return Array.isArray(results)
        ? results.map((r) => ({
            uid: r.uid,
            label: labelField.map((L) => r[L]).join(' | '),
          }))
        : [];
    }
    return Array.isArray(results)
      ? results.map((r) => ({
          uid: r.uid,
          label: r[labelField],
        }))
      : [];
  }, [labelField, results]);

  const goBackToSearchField = () => {
    setWaitForInput(true);
    searchFieldRefsCombined.current && searchFieldRefsCombined.current.select();
    setShowResults(false);
  };

  const handleListItemSelect = React.useCallback(
    (selectedId) => {
      if (!loading) {
        const selectedData = results.find((r) => r.uid == selectedId);
        if (!selectedData) throw Error('Selected item not found');
        setIsFocused(false);
        setShowResults(false);
        setSearchValue(
          selectedData[Array.isArray(labelField) ? labelField[0] : labelField]
        );
        goBackToSearchField();
        setWaitForInput(true);
        handleSelect(selectedData);
      } else {
        console.log('loading');
      }
    },
    [loading, results, goBackToSearchField, handleSelect]
  );

  const handleCancel = () => {
    setShowResults(false);
    setIsFocused(false);
    setLoading(false);
    setWaitForInput(false);
  };

  const handleInputKeyDown = (e) => {
    /* When in search field */
    if (e.key === 'Enter') {
      e.preventDefault();
      setShowResults(true);
    }
    if (e.key === 'Escape' || e.key === 'Tab') {
      handleCancel();
    }
    if (e.key === 'ArrowDown') {
      if (mappedResults && mappedResults.length > 0) {
        e.preventDefault();
        if (showResults && !loading) {
          firstItemRef.current?.focus();
        } else {
          setShowResults(true);
        }
      }
    }
  };

  const classNames = [className, 'sas_drop_wrap ', valid ? '' : 'invalid']
    .filter((a) => a)
    .join(' ');

  const handleEnterSearchField = () => {
    setIsFocused(true);
  };

  const handleClickDropdownBtn = () => {
    if (showResults) setShowResults(false);
    else {
      goBackToSearchField();
      setWaitForInput(false);
      setIsFocused(true);
    }
  };

  const handleDropdownClose = () => {
    handleCancel();
  };

  const handleInputClick = () => {
    setIsFocused(true);
    setWaitForInput(false);
  };
  return (
    <SasDropWrap className={classNames}>
      <SearchFieldWrapStyle className="searchFieldWrap">
        <input
          className="searchField"
          style={{ flexGrow: 1 }}
          type="text"
          // aria-label="Search input for selection dropdown"
          placeholder={searchFieldPlaceholder}
          value={searchValue || ''}
          onChange={onSearchFieldChange}
          onFocus={handleEnterSearchField}
          onBlur={() => setIsFocused(false)}
          onKeyDown={handleInputKeyDown}
          ref={searchFieldRefsCombined}
          onClick={handleInputClick}
          {...additionalProps}
        />
        {!loading && (
          <DropdownBtn
            type="button"
            name="Show Results"
            aria-label="Show Results"
            className="dropdownBtn"
            onClick={handleClickDropdownBtn}
          >
            \/
          </DropdownBtn>
        )}
        <SasDropLoadingWrap className="sas_drop_loadingWrap">
          <LoadingIcon isLoading={loading} />
        </SasDropLoadingWrap>
      </SearchFieldWrapStyle>
      <CustomSelectDropdown
        options={mappedResults}
        isOpen={showResults}
        handleSelect={(uid) => handleListItemSelect(uid)}
        handleClose={handleDropdownClose}
        firstItemRef={firstItemRef}
        goBackToSearchField={goBackToSearchField}
        position="relative"
      />
    </SasDropWrap>
  );
};

// TODO: prop types causing issue with ...rest

const ValueTypes = PropTypes.oneOfType([
  PropTypes.string,
  PropTypes.number,
  PropTypes.shape({
    uid: PropTypes.string,
    // TODO: Define initial value shape requirements
  }),
]);

SearchAndSelectDropdown.propTypes = {
  /* Async function to call when searching
   * Signature: async (searchText) => {}
   */
  searchFunction: PropTypes.func.isRequired,
  /* Function called when item is selected
   * Signature: (selectedId, selectedData) => {}
   */
  handleSelect: PropTypes.func.isRequired,
  /* Initial search field value */
  initialValue: PropTypes.oneOfType([
    ValueTypes,
    PropTypes.arrayOf(ValueTypes),
  ]),
  /* the target field that the search string applies to */
  searchFieldTargetField: PropTypes.string,
  /* The field in the returned data to use as the label */
  labelField: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.arrayOf(PropTypes.string),
  ]).isRequired,
  /* Additional classnames */
  className: PropTypes.string,
  /* Search field placeholder */
  searchFieldPlaceholder: PropTypes.string,
  /* Call search even when input is empty */
  allowEmptySearch: PropTypes.bool,
  /* Delay between input and search call */
  searchDelay: PropTypes.number,
  /* Input is valid */
  valid: PropTypes.bool,
  /* If true allows multiple selection */
  allowMultiple: PropTypes.bool,
};

SearchAndSelectDropdown.defaultProps = {
  searchFieldTargetField: 'label',
  className: '',
  searchFieldPlaceholder: 'search...',
  allowEmptySearch: false,
  searchDelay: 500,
  valid: true,
  initialValue: null,
  allowMultiple: false,
};
