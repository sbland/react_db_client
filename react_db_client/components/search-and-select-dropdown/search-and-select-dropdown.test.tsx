import React from 'react';
import { screen, render, within, waitForElementToBeRemoved, waitFor } from '@testing-library/react';
import UserEvent from '@testing-library/user-event';
import * as compositions from './search-and-select-dropdown.composition';
import { demoResultData } from './demo-data';

describe('SearchAndSelect', () => {
  describe('Compositions', () => {
    Object.entries(compositions).forEach(([name, Composition]) => {
      test(name, () => {
        render(<Composition />);
      });
    });
  });
  describe('loading results', () => {
    test('should not do anything when we focus on search field', async () => {
      render(<compositions.DemoData />);
      const searchField = screen.getByRole('textbox');
      await UserEvent.click(searchField);
      const resultList = screen.queryByRole('list');
      expect(resultList).not.toBeInTheDocument();
    });
    test('should get list of items when allow empty is true', async () => {
      render(<compositions.DemoDataAllowEmptyInstant />);
      const searchField = screen.getByRole('textbox');
      expect(screen.queryByRole('list')).not.toBeInTheDocument();
      await UserEvent.click(searchField);
      const resultList = await screen.findByRole('list', {}, { timeout: 3000 });
      const resultItems = await within(resultList).findAllByRole('listitem');
      expect(resultItems.length).toBeGreaterThan(0);
    });

    test('should get list of items when changing input', async () => {
      render(<compositions.DemoData />);
      const searchField = screen.getByRole('textbox');
      expect(screen.queryByRole('list')).not.toBeInTheDocument();
      await UserEvent.click(searchField);
      await UserEvent.keyboard('A');
      await screen.findByTestId('loadingIcon');
      const resultList = await screen.findByRole('list', {}, { timeout: 3000 });
      const resultItems = await within(resultList).findAllByRole('listitem');
      expect(resultItems.length).toBeGreaterThan(0);
    });
    test('should call show results when focused', async () => {
      render(<compositions.DemoData />);
      const searchField = screen.getByRole('textbox');
      await UserEvent.click(searchField);
      await UserEvent.keyboard('A');
      await screen.findByRole('list', {}, { timeout: 3000 });
      await UserEvent.click(document.body);
      await waitFor(() => expect(screen.queryByRole('list')).not.toBeInTheDocument());
      const dropdownBtn = screen.getByRole('button', { name: /Show Results/ });
      await UserEvent.click(dropdownBtn);
      const resultList = await screen.findByRole('list', {}, { timeout: 3000 });
      expect(resultList).toBeInTheDocument();
    });
  });
  describe('selecting result', () => {
    test('should set selection on click', async () => {
      render(<compositions.DemoDataAllowEmptyInstant />);
      const searchField = screen.getByRole('textbox');
      await UserEvent.click(searchField);
      const resultList = await screen.findByRole('list', {}, { timeout: 3000 });
      const resultItems = await within(resultList).findAllByRole('listitem');
      const firstItemBtn = within(resultItems[0]).getByRole('button');
      await UserEvent.click(firstItemBtn);
      expect(screen.getByTestId('curSel').textContent).toEqual(demoResultData[0].uid);
      expect(screen.queryByRole('list')).not.toBeInTheDocument();
    });
    test('should set the search value to match the selected items label field', async () => {
      render(<compositions.DemoDataAllowEmptyInstant />);
      const searchField = screen.getByRole('textbox');
      await UserEvent.click(searchField);
      const resultList = await screen.findByRole('list', {}, { timeout: 3000 });
      const resultItems = await within(resultList).findAllByRole('listitem');
      const firstItemBtn = within(resultItems[0]).getByRole('button');
      await UserEvent.click(firstItemBtn);
      const searchInput: HTMLInputElement = screen.getByRole('textbox');
      expect(searchInput.value).toEqual(demoResultData[0].label);
    });
  });
  describe('forward ref', () => {
    test.todo('should pass ref to input');
  });
});
