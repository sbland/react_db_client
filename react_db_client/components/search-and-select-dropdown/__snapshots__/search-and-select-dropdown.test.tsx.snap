// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchAndSelectDropdown shallow renders Matches Snapshot 1`] = `
"<styled.div className=\\"sas_drop_wrap \\">
  <styled.div className=\\"searchFieldWrap\\">
    <input className=\\"searchField\\" style={{...}} type=\\"text\\" placeholder=\\"helloworld\\" value=\\"\\" onChange={[Function: onSearchFieldChange]} onFocus={[Function: handleEnterSearchField]} onBlur={[Function: onBlur]} onKeyDown={[Function: handleInputKeyDown]} onClick={[Function: handleInputClick]} />
    <styled.button type=\\"button\\" name=\\"Show Results\\" aria-label=\\"Show Results\\" className=\\"dropdownBtn\\" onClick={[Function: handleClickDropdownBtn]}>
      \\\\/
    </styled.button>
    <styled.div className=\\"sas_drop_loadingWrap\\">
      <LoadingIcon isLoading={false} />
    </styled.div>
  </styled.div>
  <CustomSelectDropdown options={{...}} isOpen={false} handleSelect={[Function: handleSelect]} handleClose={[Function: handleDropdownClose]} firstItemRef={{...}} goBackToSearchField={[Function: goBackToSearchField]} position=\\"relative\\" containerRef={{...}} />
</styled.div>"
`;
